test/core.jl

```diff
@@ -3586,3 +3586,13 @@ function __f_isa_arg_1()
     length(a)
 end
 @test __f_isa_arg_1() == 1
+
+# non-terminating inference, issue #14009
+type A14009{T}; end
+A14009{T}(a::T) = A14009{T}()
+f14009(a) = ? ? f14009(A14009(a)) : a
```

Comment:
is it really necessary to use `?` as a variable here? i used to think this was was an unintentional parser "bug"...